File: src/com/googlecode/utterlyidle/ProducesMimeMatcher.scala
package org.webfabric.rest

import javax.ws.rs.Produces
import java.lang.reflect.Method
import javax.ws.rs.core.{HttpHeaders, MediaType}

class ProducesMimeMatcher(method: Method) extends Matcher[Request] {
  lazy val mimeType: String = {
    List(method.getAnnotation(classOf[Produces]), method.getDeclaringClass.getAnnotation(classOf[Produces])).filter(_ != null) match {
      case x :: xs => x.value.first
      case Nil => MediaType.WILDCARD
    }
  }

  def isMatch(request: Request): Boolean = {
    if (request.headers.contains(HttpHeaders.ACCEPT)) {
      Accept(request.headers.getValue(HttpHeaders.ACCEPT)).contains(mimeType)
    } else true
  }

  def matchQuality(request: Request): Float = {
    if (request.headers.contains(HttpHeaders.ACCEPT)) {
      Accept(request.headers.getValue(HttpHeaders.ACCEPT)).quality(mimeType)
    } else 1.0f
  }
}

File: src/com/googlecode/utterlyidle/ProducesMimeMatcher.java
package com.googlecode.utterlyidle;

import com.googlecode.totallylazy.Function1;
import com.googlecode.totallylazy.Predicate;
import com.googlecode.totallylazy.Sequence;

import static com.googlecode.utterlyidle.Accept.accept;

public class ProducesMimeMatcher implements Predicate<Request> {
    private final Sequence<String> mimeTypes;

    public ProducesMimeMatcher(Sequence<String> mimeTypes) {
        this.mimeTypes = mimeTypes;
    }

    public static Function1<Binding, Float> quality(final Request request) {
        return binding -> quality(request, binding);
    }

    public static float quality(Request request, Binding firstSignature) {
        return new ProducesMimeMatcher(firstSignature.produces()).matchQuality(request);
    }

    public boolean matches(Request request) {
       return accept(request).exists(mimeTypes);
    }

    public float matchQuality(Request request) {
       return accept(request).quality(mimeTypes);
    }

    public Sequence<String> mimeType() {
        return mimeTypes;
    }

    public static Predicate<Binding> producesMatches(final Request request) {
        return binding -> new ProducesMimeMatcher(binding.produces()).matches(request);
    }
}